src/RequestHandler.php

```diff
@@ -196,8 +225,13 @@ public function slaveConnected(ConnectionInterface $connection)
         // update slave availability
         $this->connection->on('close', [$this, 'slaveClosed']);
 
+        // keep track of the last sent data to detect if slave exited abnormally
+        $this->connection->on('data', function ($data) {
+            $this->lastOutgoingData = $data;
+        });
+
         // relay data to client
-        $this->connection->pipe($this->incoming);
+        $this->connection->pipe($this->incoming, array('end' => false));
```

Comment:
What does the end do?

Reply:
By default when the connection socket closes, it pipes this event through to incoming socket and it closes as well (see https://github.com/reactphp/stream#pipe), so the 502 response would not get to the client. The end => false parameter prevents this behavior and should have no side effect since incoming->end is called in the slaveClosed function anyway.